File: 7.12.6/gwan/0.0.0.0:80/#0.0.0.0/old.csp/loan.java
// ============================================================================
// Java servlet example for the G-WAN Web App. Server (http://trustleap.ch/)
// ----------------------------------------------------------------------------
// loan.java: G-WAN's version of the Java loan http://gwan.ch/source/loan.jsp
//            written for GlassFish (follow the link above for details).
// ----------------------------------------------------------------------------
// This Java code is served very quickly by G-WAN (despite C/Java conversions). 
// See below (the " -t 6" option setups 6 client worker threads for 6 Cores):
// 
// weighttp -n 1000000 -c 300 -t 6 -k -H "Accept-Encoding: gzip" \
// "http://127.0.0.1:80/?loan.java&name=Eva&amount=10000&rate=3.5&term=1"
// finished in 1 sec, 748 millisec, 571778 req/s, 1286500 kbyte/s
// requests: 1000000 total/started/done/succeeded, 0 failed/errored
// status codes: 1000000 2xx, 0 3xx, 0 4xx, 0 5xx
// traffic: 2303999915 bytes total, 276999915 bytes http, 2027000000 bytes data
//
// On the same machine, Tomcat's HelloWorld example does:
// (installation: https://help.ubuntu.com/11.04/serverguide/C/tomcat.html)
// (administration https://help.ubuntu.com/community/ApacheTomcat5)
// 
// weighttp -n 1000000 -c 300 -t 6 -k -H "Accept-Encoding: gzip" \
// "http://127.0.0.1:8080/examples/servlets/servlet/HelloWorldExample"
// finished in 12 sec, 278 millisec, 81441 req/s, 39024 kbyte/s
// requests: 1000000 total/started/done/succeeded, 0 failed/errored
// status codes: 1000000 2xx, 0 3xx, 0 4xx, 0 5xx
// traffic: 490671704 bytes total, 131671704 bytes http, 359000000 bytes data
// 
//   Server    bytes sent    kbyte/sec  requests/sec   servlet     time
//   ------  --------------  ---------  ------------  ----------  ------
//   G-WAN    2,303,999,915  1,286,500       571,778   loan.java   1 sec
//   Tomcat     490,671,704     39,024        81,441  hello.java  12 sec
// 
// The next best Web servers do not even reach 571,778 req/s when serving mere
// static contents. G-WAN is making dynamic contents fly (high).
// ----------------------------------------------------------------------------
import api.Gwan;

public class loan // the class name MUST match the script *.java filename
{
   public static int jmain(long env, String[] args) 
   {
      String m_out = "";

      // ---- requests without parameters trigger a redirect to the input form
      if(args.length == 0)
      {
         m_out = "HTTP/1.1 302 Found\r\n"
            + "Content-type:text/html\r\n"
            + "Location: /csp_loan.html\r\n\r\n"
            + "<!DOCTYPE HTML PUBLIC \"-//IETF//DTD HTML 2.0//EN\">"
            + "<html><head><title>Redirect</title></head>"
            + "<body>Click <a href=\"/csp_loan.html\">HERE</A> for redirect.</body>"
            + "</html>";
      }
      else
      {
         // ---- process a GET query with parameters
         long start = Gwan.getUs();
	      double payment, interest, principal, cost, amount, rate, term;
	      int    month=0, year=1, lastpayment=1;
         String[] Months={"January","February","March","April","May","June",
                  "July", "August","September","October","November","December"};

         // the form field "names" we want to find values for
         String szName="-", Name="", Amount="0.0", Rate="0.0", Term="0.0";

         // anyone volunter to write a Java equivalent for G-WAN's get_arg()?
         switch(args[0].charAt(0))
         {
             case 'a': Amount = args[0].substring(args[0].indexOf('=') + 1); break;
             case 'n': Name   = args[0].substring(args[0].indexOf('=') + 1); break;
             case 'r': Rate   = args[0].substring(args[0].indexOf('=') + 1); break;
             case 't': Term   = args[0].substring(args[0].indexOf('=') + 1); break;
         }
         switch(args[1].charAt(0))
         {
             case 'a': Amount = args[1].substring(args[1].indexOf('=') + 1); break;
             case 'n': Name   = args[1].substring(args[1].indexOf('=') + 1); break;
             case 'r': Rate   = args[1].substring(args[1].indexOf('=') + 1); break;
             case 't': Term   = args[1].substring(args[1].indexOf('=') + 1); break;
         }
         switch(args[2].charAt(0))
         {
             case 'a': Amount = args[2].substring(args[2].indexOf('=') + 1); break;
             case 'n': Name   = args[2].substring(args[2].indexOf('=') + 1); break;
             case 'r': Rate   = args[2].substring(args[2].indexOf('=') + 1); break;
             case 't': Term   = args[2].substring(args[2].indexOf('=') + 1); break;
         }
         switch(args[3].charAt(0))
         {
             case 'a': Amount = args[3].substring(args[3].indexOf('=') + 1); break;
             case 'n': Name   = args[3].substring(args[3].indexOf('=') + 1); break;
             case 'r': Rate   = args[3].substring(args[3].indexOf('=') + 1); break;
             case 't': Term   = args[3].substring(args[3].indexOf('=') + 1); break;
         }

         // set default values if a parameter is missing
         if(Name   == null) Name   ="-";
         if(Amount == null) Amount ="100000.0";
         if(Rate   == null) Rate   ="3.5";
         if(Term   == null) Term   ="10.0";

         // all litteral strings provided by a client must be escaped this way
         // if you inject them into an HTML page
         //public static String escape_html(String Name)
         {
             int len = Name.length();
             StringBuffer sb = new StringBuffer(len);
             int c;

             for(int i = 0; i < len; i++)
             {
                 c = Name.charAt(i);
                 if(c == ' ')  sb.append("&#32;");  else
                 if(c == '"')  sb.append("&quot;"); else
                 if(c == '&')  sb.append("&amp;");  else
                 if(c == '<')  sb.append("&lt;");   else
                 if(c == '>')  sb.append("&gt;");   else
                 if(c == '\n') sb.append("&lt;br/&gt;");
                 else
                 {
                    c = c&0xffff; // unicode
                    if(c < 32 || c > 127)
                    {
                       sb.append("&#");
                       sb.append(new Integer(c).toString());
                       sb.append(';');
                    }
                    else
                       sb.append(c);
                 }
             }
             //return sb.toString();
             szName = sb.toString();
         }

         // filter input data to avoid all the useless or nasty cases
         amount = Double.parseDouble(Amount);
         if(amount < 1.0)
            amount = 1.0;

         rate = Double.parseDouble(Rate);
         if(rate > 19.0)
            rate = 20.0;

         if(rate > 1.0)
            rate = rate/100.0;
         else
         if(rate < 1.0)
            rate  = 1.0/100.0;

         term = Double.parseDouble(Term);
	      if(term < 0.1)
	         term = 1.0/12.0;
         // do not do this in production...
	      //else
	      //if(term > 90.0)
         //   term = 90.0;

         // calculate the monthly payment amount
         payment = amount*rate/12*Math.pow(1+rate/12, term*12)
                 /               (Math.pow(1+rate/12, term*12)-1);
         cost    = (term*12*payment)-amount;

         // build the top of our HTML page
         m_out += String.format(
           "<!DOCTYPE HTML>"
         + "<html lang=\"en\"><head><title>Loan Calculator</title><meta http-equiv"
         + "=\"Content-Type\" content=\"text/html; charset=utf-8\">"
         + "<link href=\"/imgs/style.css\" rel=\"stylesheet\" type=\"text/css\">"
         + "</head><body style=\"margin:16px;\">"
         + "<h2>Dear %s, your loan goes as follows:</h2><br>", 
         (Name.length() < 2 || Name.charAt(0) == '-') ? "client" : Name);
         // the HTML escaping needs a fix, any volunteer?
         //(szName.length() < 2 || szName.charAt(0) == '-') ? "client" : szName);

         m_out += String.format(
                    "<table class=\"clean\" width=240px>"
                   +"<tr><th>loan</th><th>details</th></tr>"
                   +"<tr class=\"d1\"><td>Amount</td><td>%,.2f"
                   +"</td></tr><tr class=\"d0\"><td>Rate</td><td>"
                   +"%,.2f%%</td></tr><tr class=\"d1\"><td>Term</td><td>"
                   +"%d %s(s)</td></tr><tr class=\"d0\"><td>Cost</td><td>"
                   +"%,.2f (%3.2f%%)</td></tr></table>",
                    amount,
                    rate * 100,
                    (int)(((int)term>0)?term:Math.ceil(12*term)),
                    ((int)term>0)?"year":"month",
                    cost,
                    100/(amount/cost));

	      m_out += String.format(
	                 "<br><table class=\"clean\" width=112px>"
                   +"<tr class=\"d1\"><td><b>YEAR %d"
                   +"</b></td></tr></table>"
                   +"<table class=\"clean\" width=550px>"
                   +"<tr><th>month</th><th>payment</th><th>interest</th>"
                   +"<th>principal</th><th>balance</th></tr>",
                    year);

         // output monthly payments
	      for(;;)
	      {
		      month    = month+1;
      	       interest = (amount*rate)/12;
		      if(amount>payment)
            {
               amount    = (amount-payment)+interest;
		         principal = payment-interest;
            }
            else
            {
               // calculate last payment
               if(lastpayment>0)
               {
                  lastpayment = 0;
                  payment     = amount;
                  principal   = amount-interest;
                  amount      = 0;
               }
               else
               {
                  // all payments are done, just padd the table
                  amount    = 0;
                  payment   = 0;
                  interest  = 0;
                  principal = 0;
               }
            }

            m_out += String.format(
                         "<tr class=\"d%d\"><td>%s</td><td>%,.2f</td><td>%,.2f"
                        +"</td><td>%,.2f</td><td>%,.2f</td></tr>",
                         month & 1, Months[month - 1],
                         payment,
                         interest,
                         principal,
                         amount);

		      if(month==12)
		      {
               if(amount>0)
               {
			         month=0; year=year+1;
	               m_out += String.format(
	                            "</table><br><table class=\"clean\" width=112px>"
                              +"<tr class=\"d1\"><td><b>YEAR %d"
                              +"</b></td></tr></table>"
                              +"<table class=\"clean\" width=550px>"
                              +"<tr><th>month</th><th>payment</th><th>interest"
                              +"</th><th>principal</th><th>balance</th></tr>",
                               year);
               }
               else
               {
                  break;
               }
            }

         }

         // time the process and close the HTML page
         m_out += String.format(
                    "</table><br>This page was generated in %,.2f ms."
                   +"<br>(on a 3GHz CPU 1 ms = 3,000,000 cycles)"
                   +"<br></body></html>", (Gwan.getUs() - start)/1000.0);

      }

      Gwan.xbufCat(Gwan.getReply(env), m_out);
      return 200;
   }
}
// ============================================================================
// End of Source Code
// ============================================================================

File: 7.12.6/gwan/0.0.0.0:80/#0.0.0.0/old.csp/loan.php
<?php
  /* loan.php: A loan calculator written in PHP 5.4.9.1 (procedural mode)
     compatible with previous  5.2.x and 5.3.x PHP stacks
     to compare with loan.c, loan.cs and loan.java
     Must be launched with 4 parameters such as :
     loan.php&name=Eva&amount=10000&rate=3.5&term=1
  */
  $time_start = microtime(true);
	
  /* convert PHP's command-line arguments into PHP's URI parameters */
  if(PHP_SAPI==="cli") 
  {
     parse_str(implode('&', array_slice($argv, 1)), $_GET);
  }
  /* if we have the 4 query parameters, we process a GET/POST form */
  if ((isset($_GET['name']))&&(isset($_GET['amount']))&&(isset($_GET['rate']))&&(isset($_GET['term']))) 
  {
    $name = $_GET['name']; 
  }
  else 
  { /* we redirect to the /csp_loan.html page */
    ?>
    <!DOCTYPE HTML>
      <html>
        <head>
          <meta http-equiv='Content-Type' content='text/html; charset=UTF-8' />
          <title>Loan calculator</title>
          <link type="text/css" rel="stylesheet" href="/www/imgs/style.css"></link>
        </head>
        <body>
          <script language="javascript" type="text/javascript">
            function load_Page(){parent.window.location='/csp_loan.html';}
            window.onload=function(){load_Page();}
          </script>
        </body>
      </html>
    <?php
  }

  if($name == '' || $name == '-' || $name == ' ') 
  {
    $name = "client";
  }

  if($name != '') 
  {
    $monthList = array("January", 
                       "February", 
                       "March", 
                       "April", 
                       "May", 
                       "June", 
                       "July", 
                       "August", 
                       "September", 
                       "October",
                       "November", 
                       "December", 
                       "");

    if (isset($_GET['amount'])) 
    {
      $amount = floatval(trim($_GET['amount'])); 
    }
       
    if ($amount == '' || $amount == 0 || $amount == ' ') 
    {
      $amount = 100000.00;
    }

    if($amount < 1) 
    {
      $amount = 100000.00;
    }

    if (isset($_GET['rate'])) 
    {
      $rate = floatval(trim($_GET['rate']));
    }

    if ($rate == '' || $rate == 0 || $rate == ' ') 
    {
      $rate = 3.5;
    }
    else 
    {
      $rate = abs($rate);
    }

    if($rate > 20) /* to limit the damage - LoL */
    { 
      $rate = 20;
    }

    if($rate > 1) 
    {
      $rate = abs($rate / 100.);
    }
    else 
    {
      $rate = 3.5 / 100.;
    }

    if (isset($_GET['term'])) 
    {
      $term = floatval(trim($_GET['term']));
    }

    if ($term == '' || $term < 1 || $term == ' ') 
    {
      $term = 1;
    }
    elseif ($term > 800) /* to limit the damage - LoL */
    {
      $term = 800;
    }
    else 
    {
      $term = round(abs($term));
    }

    $name = addslashes($name);
    $year = 1; 
    $lastpayment = 1;
    /* all literal strings provided by a client must be escaped this way
     * if you inject them into an HTML page
     */
    htmlspecialchars($name);
    /* calculate the montly payment amount */
    $payment = $amount*$rate/12*pow(1+$rate/12,$term*12)/(pow(1+$rate/12,$term*12)-1);
    /* build the top of our HTML page */
    $cost = ($term*12*$payment)-$amount;
    $result  = "<!DOCTYPE html PUBLIC \"-//W3C//DTD XHTML 1.0 Transitional//EN\" \"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\">";
    $result .= sprintf("<html>" .
                       "<head>" .
                       "<meta http-equiv='Content-Type' content='text/html; charset=UTF-8' />" .
                       "<title>Loan calculator</title>" .
                       "<link type=\"text/css\" rel=\"stylesheet\" href=\"imgs/style.css\"></link>" .
                       "</head>" .
                       "<body>" .
                       "<h2>Dear $name, your loan goes as follows:</h2>");
    $result .= sprintf("<br/><table class=\"clean\" width=240px>" .
                       "<tr><th>loan</th><th>details</th></tr>" .
                       "<tr class=\"d1\"><td>Amount</td><td>%s</td></tr>" .
                       "<tr class=\"d0\"><td>Rate</td><td>%.2f%%</td></tr>" .
                       "<tr class=\"d1\"><td>Term</td><td>%u %s(s)</td></tr>" .
                       "<tr class=\"d0\"><td>Cost</td><td>%s (%.2f%%)</td></tr>" .
                       "</table>", number_format($amount, 2), $rate*100, 
                       $term,($term>0)?"year":"month", 
                       number_format($cost,2), 100/($amount/$cost));
                       
    $result .= sprintf("<br/><table class=\"clean\" width=112px>" .
                       "<tr class=\"d1\"><td><b>YEAR %u</b></td></tr></table>" .
                       "<table class=\"clean\" width=550px>" .
                       "<tr><th>month</th><th>payment</th><th>interest</th>" .
                       "<th>principal</th><th>balance</th></tr>", $year);
                       
    for($month = 1; $month <= 12; $month++) // ouput monthly payments
    { 
      $interest = ($amount*$rate)/12;
      if($amount > $payment) 
      {
        $amount = ($amount-$payment) + $interest; 
        $principal = $payment - $interest;
      }
      else // calculate last payment
      { 
        if($lastpayment > 0) 
        {
          $lastpayment = 0;
          $payment = $amount;
          $principal = $amount - $interest;
          $amount = 0;
        }
        else // all payments are done, just pad the table
        { 
          $amount = 0;
          $payment = 0;
          $interest = 0;
          $principal = 0;
        }                
      }
      $result .= sprintf("<tr class=\"d%u\"><td>%s</td><td>%s</td><td>" .
                         "%s</td><td>%s</td><td>%s</td></tr>",
                         $month%2, $monthList[$month-1], 
                         number_format($payment, 2), 
                         number_format($interest, 2), 
                         number_format($principal, 2), 
                         number_format($amount, 2));
      if($month == 12) 
      {
        if($amount) 
        {
          $month = 0; 
          $year++;
          $result .= sprintf("</table><br><table class=\"clean\" " .
                             "width=112px><tr class=\"d1\"><td><b>YEAR %u</b>" .
                             "</td></tr></table><table class=\"clean\" width=550px>" .
                             "<tr><th>month</th><th>payment</th><th>interest</th>" .
                             "<th>principal</th><th>balance</th></tr>",
                             $year);
        }
      }
    }
    /* time the process, print the HTML page and close the program with a "200" return code */
    $result .= sprintf("</table><br>This page was generated in %.2F ms.</br>(on a 3GHz CPU 1 ms = 3,000,000 cycles)".
    "</br></body></html>",number_format(((microtime(true)-$time_start)*1000),2));
    echo $result;
    exit(200);
  }
?>

